### ISSUE TITLE
Blurring at right edge of screen

### ISSUE THREAD

**Issue by skalyan91** [Raymo111/i3lock-color#103]
When I use `i3lock` with the `--blur` option, the left edge of the screen looks normal, but the right edge has a whitish band. I assume this is because the algorithm for computing the blur at the right edge is referring to pixels beyond the right edge of the image; and so I tried changing lines 117­–121 of `blur.c` to
```
} else if (rightBorder) {
                for (; i < column - width + HALF_KERNEL; i++)
                    rgbaIn[i] = *(src - (HALF_KERNEL - i));
                for (; i < KERNEL_SIZE; i++)
                    rgbaIn[i] = *(src + (HALF_KERNEL - i));
```
and recompiling. However, this has no effect.

A similar issue exists at the bottom of the screen (as opposed to the top), and I assume the cause is similar.

**Comment by PandorasFox** [comment]
What resolution are you running at, and what CPU do you have? This is dependent upon which instructions your CPU has available + the specific res that this occurs on.

**Comment by skalyan91** [comment]
My screen resolution is 3200×1800, and I have an Intel Core i7 CPU.

**Comment by PandorasFox** [comment]
What generation CPU? What specific model? Older i7s don't always support the instruction extensions used in some of the blurring code.

**Comment by skalyan91** [comment]
i7-7560U.

**Comment by PandorasFox** [comment]
You'll need to look [here](https://github.com/PandorasFox/i3lock-color/blob/master/blur_simd.c) then. Your CPU has SSE2 support, so the changes you're making to the generic unaccelerated passes won't do anything.

**Comment by skalyan91** [comment]
Aha! Thanks; I’ll play with that, and submit a pull request if I manage to work it out.

**Comment by skalyan91** [comment]
Oh, wait a sec—do the pixels wrap around for the purposes of computing the blur? That would account for my issue, as the left side of my wallpaper has a light colour, whereas the right side has a dark colour (and likewise the top is light, and the bottom is dark). Playing around with a few different wallpapers suggests that this is indeed the case.

Is there a way you could get it to not treat the pixels as wrapping around, but just, say, treat the edge pixels as stretching out to infinity? I know almost no C, so I wouldn’t be confident about making these kinds of changes to the code on my own.

**Comment by skalyan91** [comment]
FWIW, the blur function in ImageMagick shows the desired behaviour. Maybe just add an option to do the blurring using ImageMagick?

**Comment by PandorasFox** [comment]
For the former - probably, but it complicates things a lot, since how to blur it isn't as well defined.

For the latter, you can do that already and just pass in the image as an arg like normal.

**Comment by skalyan91** [comment]
I’ve now worked out how to use ImageMagick and pass the result to the `-i` argument of `i3lock`. However, your blurring is _way_ faster than the ImageMagick blurring. For now, I’ve decided to just save a blurred version of my wallpaper and use that for the lockscreen, rather than blurring a screenshot.

I think you should still consider addressing this wrapping issue, though. The problem must be well-defined if ImageMagick (and apparently Photoshop) solve it.

**Comment by meskarune** [comment]
The problem has not already been solved by others, as other solutions are incredibly slow. You can have a blur that is fast and works for most people, or one that is slow but precise like using a script with image magick. (https://github.com/meskarune/i3lock-fancy is one example of a script that looks nice but runs significantly slower than i3lock-color --blur 5) I don't know that its possible to have both speed and perfection at once.

@skalyan91 it is also possible for image magick to create white lines on the side of the screen due to rounding errors when blurring, I actually had this issue with some monitor resolutions. The fix was to use a screenshot that was slightly larger than the actual resolution.

If the white line issue is caused by a rounding error in i3lock-color then maybe the code could also simply assume an image size slightly larger than the resolution. Like by just a few pixels might be enough.

**Comment by PandorasFox** [comment]
I think the problem with that approach is just that mostly, it's kinda hard to really do that. That'll mean allocating more memory / figuring out how much is 'enough' / incurring a performance cost / etc. I'm not really sure it would be doable, unless I just mess around with the blur pixmap.

In either case - yeah, the point of the blurring in i3lock-color is just so that instead of doing something like `feh | imagemagick > /tmp/blur.png && i3lock -i /tmp/blur.png`, you can just do `i3lock -B 5` and not incur the performance overhead of all the other stuff / have i3lock just immediately grab the screen.

Basically, the point of the screen locker is to be fast and secure (and this fork also adds "pretty" to the mix). Bringing it all in-house makes the process faster, and it means that a problem in feh or imagemagick doesn't prevent i3lock from running.

I also don't want to make the code more complicated than it needs to be. I'm not sure there'd be a 'good' solution to this I'd take.
